File: sdk/go/integration_test.go
//go:build integration

// Integration test for the SDK against a running API. Run with:
//
//	go test -tags integration ./... -run TestSDKLive -v
//
// Override the base URL with AETERNISLOG_BASE_URL (default http://localhost:5001).
package aeternislog

import (
	"context"
	"os"
	"testing"
	"time"
)

func TestSDKLive(t *testing.T) {
	base := os.Getenv("AETERNISLOG_BASE_URL")
	if base == "" {
		base = "http://localhost:5001"
	}
	c := New(base, WithAPIKey(os.Getenv("AETERNISLOG_API_KEY")))
	ctx := context.Background()
	const domain = "sdktest"

	// Create a record; CreateRecord checks the server hash == the local hash.
	rec, err := c.CreateRecord(ctx, domain, "sdk", map[string]interface{}{"foo": "bar", "n": 7}, nil)
	if err != nil {
		t.Fatalf("create: %v", err)
	}
	t.Logf("created %s hash=%s", rec.ID, rec.Hash)

	// Fetch it back; its independently recomputed hash must match the stored one.
	got, err := c.GetRecord(ctx, domain, rec.ID)
	if err != nil {
		t.Fatalf("get: %v", err)
	}
	if got.ComputeHash() != got.Hash {
		t.Errorf("local hash %s != stored hash %s", got.ComputeHash(), got.Hash)
	}

	// Batch + anchor, then verify server-side.
	time.Sleep(150 * time.Millisecond)
	res, err := c.BatchRecords(ctx, domain)
	if err != nil {
		t.Fatalf("batch: %v", err)
	}
	if !res.Anchored || res.TxID == "" {
		t.Fatalf("expected an anchored batch with a tx id, got %+v", res)
	}
	t.Logf("batch %s anchored root=%s tx=%s", res.BatchID, res.MerkleRoot, res.TxID)

	v, err := c.VerifyBatch(ctx, domain, res.BatchID)
	if err != nil {
		t.Fatalf("verify: %v", err)
	}
	if !v.IsValid {
		t.Errorf("expected VALID, got %+v", v)
	}
}

File: sdk/python/tests/integration_test.py
"""Live integration test against a running API. Not auto-discovered (filename does
not match test*.py). Run explicitly:

    AETERNISLOG_BASE_URL=http://localhost:5001 \\
      python -m unittest tests.integration_test -v

Set AETERNISLOG_API_KEY too when the API has auth enabled.
"""
import os
import time
import unittest

from aeternislog import Client


@unittest.skipUnless(os.getenv("AETERNISLOG_BASE_URL"), "set AETERNISLOG_BASE_URL to run the live test")
class TestSDKLive(unittest.TestCase):
    def setUp(self):
        self.client = Client(os.environ["AETERNISLOG_BASE_URL"], api_key=os.getenv("AETERNISLOG_API_KEY", ""))
        self.domain = "pysdktest"

    def test_create_batch_verify(self):
        # create_record raises if the server hash != the locally computed hash, so a
        # pass here already proves Python<->Go canonicalization parity.
        rec = self.client.create_record(self.domain, "pysdk", {"foo": "bar", "n": 7})
        self.assertEqual(rec.hash, rec.compute_hash())

        # Fetch it back; its independently recomputed hash must match the stored one.
        got = self.client.get_record(self.domain, rec.id)
        self.assertEqual(got.compute_hash(), got.hash)

        # Batch + anchor on the blockchain, then verify server-side.
        time.sleep(0.2)
        batch = self.client.batch_records(self.domain)
        self.assertTrue(batch.anchored, batch)
        self.assertTrue(batch.tx_id, "expected a real transaction id")

        result = self.client.verify_batch(self.domain, batch.batch_id)
        self.assertTrue(result.is_valid, result)


if __name__ == "__main__":
    unittest.main()
